src/cargo/ops/cargo_new.rs

```diff
@@ -925,3 +933,103 @@ fn update_manifest_with_inherited_workspace_package_keys(
+    let workspace_root = root_manifest_path.parent().with_context(|| {
+        format!(
+            "workspace root manifest doesn't have a parent directory `{}`",
+            root_manifest_path.display()
+        )
+    })?;
+    let relpath = pathdiff::diff_paths(package_path, workspace_root).with_context(|| {
+        format!(
+            "path comparison requires two absolute paths; package_path: `{}`, workspace_path: `{}`",
+            package_path.display(),
+            workspace_root.display()
+        )
+    })?;
+
+    // Don't add the new package to the workspace's members
+    // if there is an exclusion match for it.
+    if let Some(members) = workspace_document
+        .get("workspace")
+        .and_then(|workspace| workspace.get("exclude"))
+        .and_then(|members| members.as_array())
+    {
+        for member in members {
+            let pat = member
+                .as_str()
+                .with_context(|| format!("invalid members' element `{}`", member))?;
+            if Path::new(pat) == relpath {
+                return Ok(());
+            }
+        }
+    }
+
+    let display_path = relpath
+        .to_str()
+        .map(|s| s.replace('\\', "/"))
+        .with_context(|| format!("invalid non-unicode path `{}`", relpath.display()))?;
+
+    // If the members element already exist, check if one of the patterns
+    // in the array already includes the new package's relative path.
+    // - Add the relative path if the members don't match the new package's path.
+    // - Create a new members array if there are no members element in the workspace yet.
+    if let Some(members) = workspace_document
+        .get("workspace")
+        .and_then(|workspace| workspace.get("members"))
+        .and_then(|members| members.as_array())
+    {
+        for member in members {
+            let pat = member.as_str().unwrap();
+            let pattern = glob::Pattern::new(pat)
+                .with_context(|| format!("cannot build glob pattern from `{}`", pat))?;
+
+            if pattern.matches_path(&relpath) {
+                return Ok(());
+            }
+        }
+
+        let Some(toml_members) = workspace_document["workspace"]["members"].as_array_mut() else {
+            anyhow::bail!("[workspace.members] is not an array");
+        };
+
+        toml_members.push(&display_path);
+    } else if workspace_document.contains_key("workspace") {
+        let ws = workspace_document["workspace"]
+            .as_table_mut()
+            .with_context(|| format!("[workspace] is not a table"))?;
+
+        let mut array = Array::new();
+        array.push(&display_path);
+
+        // only add a trailing new line if the manifest doesn't
+        // include any elements that are not inline tables.
+        // This respects the previous document formatting if there was any already.
+        if !ws.get_values().iter().any(|(_, v)| !v.is_inline_table()) {
```

Comment:
Why does something being an inline table matter to this check?  Would we also need to remove a previous trailing newline?